### PR TITLE
Minor Docker Optimizations

### PR DESCRIPTION
**Pull request by johnwesley**
## Description

Provide a `Makefile` to streamline the overall Docker process.
This includes an image build, container deploy, and cleanup target.

Remove `virtualenv` activation from the image build process. Containers are ephemeral in nature and I don't think we need the extra layer of virtualization in this case, however, `algo` inherently depends on it; so we still need to install the package and set the `env` for deploy.  

## Motivation and Context

This change is not required, but I believe the `Makefile`  adds a level of convenience for Docker users.
Removing `virtualenv` activation from the build process installs `ansible` to the standard `/usr/local/bin` path so it's always available. I've had issues with this in the past with algo not finding the `ansible-playbook` binary location. 

## How Has This Been Tested?

This has only been tested with `macOS 10.14.3`, `docker 18+`, and `DigitalOcean`. I can provide further tests in other environments if the maintainers are are open to these changes. This change will also require some documentation additions.

I have deployed with both `docker-deploy` Makefile target and `docker run --cap-drop=all -it -v ${HOME}/VPNS:/data trailofbits/algo:latest`, pointing to a local copy for `config.cfg`.

### Caveats
1. `./algo-showenv.sh: line 72: ./env/bin/python: cannot execute binary file: Exec format error` is reported by `algo` because we have circumvented this.

## Types of changes

- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [] I have updated the documentation accordingly.
- [] I have added tests to cover my changes.
- [] All new and existing tests passed.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -15,11 +15,11 @@ RUN mkdir -p /algo && mkdir -p /algo/configs
 
 WORKDIR /algo
 COPY requirements.txt .
+
 RUN apk --no-cache add ${BUILD_PACKAGES} && \
     python -m pip --no-cache-dir install -U pip && \
     python -m pip --no-cache-dir install virtualenv && \
     python -m virtualenv env && \
```
Comment (jackivanov) [inline Dockerfile]: This one is should be removed as well

**Comment by jackivanov** [comment]
Let's leave travis-ci for another issue and just delete `python -m virtualenv env` from the docker file

**Comment by johnwesley** [comment]
I've about finished up, just one outstanding issue with `update-users.sh`. Can I open another PR with my working branch? https://travis-ci.org/johnwesley/algo

And should we keep the `Makefile` in this branch?

**Comment by jackivanov** [comment]
Oh, if you have it already, lets address it in this PR as well :) Yes, keep Makefile in the same branch

**Comment by johnwesley** [comment]
Merged the branch I've been testing against travis-ci. 

Test `local-deploy.sh` appears to pass.

There is still an issue with `update-users.sh` to resolve.

```
TASK [Generate the SSH private key] ********************************************
fatal: [10.0.8.24]: FAILED! => {"changed": false, "msg": "the python pyOpenSSL module is required"}
```

**Comment by johnwesley** [comment]
@jackivanov  All tests now pass.  https://travis-ci.org/johnwesley/algo

**Comment by johnwesley** [comment]
@jackivanov have you had a chance to review? And can I move forward with documentation updates?

**Comment by jackivanov** [comment]
@johnwesley After my recent changes you would probably need to rebase from the master, and you can update the docs as well

**Comment by johnwesley** [comment]
Rebased with upstream master.
https://travis-ci.org/johnwesley/algo

Moving on to documentation.

**Comment by jackivanov** [comment]
Looks like the branch is messed up with non related commints now.

**Comment by johnwesley** [comment]
Will reset branch with master, add `Makefile` changes, and updates testing in a separate branch.

**Comment by johnwesley** [comment]
@jackivanov I've updated the branch.

File: Dockerfile [hunk 2]
```diff
@@ -15,11 +15,11 @@ RUN mkdir -p /algo && mkdir -p /algo/configs
 
 WORKDIR /algo
 COPY requirements.txt .
+
 RUN apk --no-cache add ${BUILD_PACKAGES} && \
     python -m pip --no-cache-dir install -U pip && \
-    python -m pip --no-cache-dir install virtualenv && \
-    python -m virtualenv env && \
-    source env/bin/activate && \
+    #python -m pip --no-cache-dir install virtualenv && \
```
Comment (jackivanov) [inline Dockerfile]: probably needs to be enabled?

File: deploy-from-docker.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
+# Docker Support
```
Comment (jackivanov) [inline deploy-from-docker.md]: seems like you've overwritten the whole file somehow. Could you include only your changes?

File: tests/local-deploy.sh [hunk 4]
```diff
@@ -6,7 +6,7 @@ DEPLOY_ARGS="provider=local server=$LXC_IP ssh_user=ubuntu endpoint=$LXC_IP appa
 
 if [ "${LXC_NAME}" == "docker" ]
 then
-  docker run -it -v $(pwd)/config.cfg:/algo/config.cfg -v ~/.ssh:/root/.ssh -v $(pwd)/configs:/algo/configs -e "DEPLOY_ARGS=${DEPLOY_ARGS}" travis/algo /bin/sh -c "chown -R 0:0 /root/.ssh && source env/bin/activate && ansible-playbook main.yml -e \"${DEPLOY_ARGS}\" --skip-tags apparmor"
+  make docker-test-local
```
Comment (jackivanov) [inline tests/local-deploy.sh]: `docker-test-local` seems to be missing in the Makefile

File: Dockerfile (hunk 2 above)
Comment (johnwesley) [reply Dockerfile]: Local and CI tests pass without `virtualenv`.

File: deploy-from-docker.md (hunk 3 above)
Comment (johnwesley) [reply deploy-from-docker.md]: Of course. Sorry.

File: tests/local-deploy.sh (hunk 4 above)
Comment (johnwesley) [reply tests/local-deploy.sh]: All tests are up to date.

**Comment by johnwesley** [comment]
@jackivanov I've (hopefully) saved this branch and addressed your concerns. Tests have been updated to use the `Makefile` and the original doc has been restored.

File: Dockerfile (hunk 2 above)
Comment (jackivanov) [reply Dockerfile]: Ah, yes, right, we agreed to run it without virtualenv. Could you then just remote those lines, please?

**Comment by jackivanov** [comment]
My tests failed:
```
root@ip-172-26-13-231:~/docker-optimize# make docker-deploy
docker run \
--cap-drop=all \
--rm \
-it \
-v /home/ubuntu/docker-optimize:/data \
trailofbits/algo:latest
/algo/env/bin/activate not found.  Did you follow documentation to install dependencies?
Makefile:20: recipe for target 'docker-deploy' failed
make: *** [docker-deploy] Error 1
```

I guess we need to modify `./algo` to skip this check if deploying from docker

**Comment by johnwesley** [comment]
This is my local play

```
$ make docker-deploy
docker run \
        --cap-drop=all \
        --rm \
        -it \
        -v /Users/kwisatz4/Github/algo:/data \
        trailofbits/algo:latest

PLAY [localhost] ******************************************************************************************************

TASK [Gathering Facts] *******************************************************************************************************
ok: [localhost]

TASK [Verify Ansible meets Drupal VM's version requirements.] *******************************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

PLAY [Ask user for the input] 
```
Will try from different environments.

**Comment by johnwesley** [comment]
And this is my local play (Arch)
```
[vagrant@archlinux algo]$ sudo make docker-deploy                                                                                                                                    │
docker run \                                                                                                                                                                         │
--cap-drop=all \                                                                                                                                                                     │
--rm \                                                                                                                                                                               │
-it \                                                                                                                                                                                │
-v /home/vagrant/algo:/data \                                                                                                                                                        │
trailofbits/algo:latest                                                                                                                                                              │
/algo/env/bin/activate not found.  Did you follow documentation to install dependencies?                                                                                             │
make: *** [Makefile:20: docker-deploy] Error 1
```

**Comment by johnwesley** [comment]
@jackivanov testing a modification to `./algo` in clean Linux and OSX environments.

**Comment by johnwesley** [comment]
Tested the modified condition with an Ubuntu 18 Vagrant environment. Both docker and standard deploys kicked off.

**Comment by dguido** [comment]
What do we need to get this merged?

**Comment by johnwesley** [comment]
Sorry, I was waiting on your feedback after my last update.
I would suggest we resolve the conflicts and the the `update-users` test. Did you have a chance to review the changes to the main `algo` script?

**Comment by jackivanov** [comment]
The fix in `algo` looks good to me. Could you, please, fix the conflicts and I'll take it over from there?

**Comment by johnwesley** [comment]
Yes I will.

File: docs/deploy-from-docker.md [hunk 5]
```diff
@@ -50,6 +50,15 @@ $ docker run --cap-drop=all -it \
 
 You can use the Dockerfile provided in this repository as-is, or modify it to suit your needs. Further instructions on building an image can be found in the [Docker engine](https://docs.docker.com/engine/) documents.
 
+## Deploying an Algo Server with a Docker Makefile
+
+A `Makefile` is included as a convenience for Docker users which aims to simplify some of the command syntax. 
+**This has not been tested in a Windows environment.**
+
+1. From the project's root run `make docker-build`. This builds a Docker image labeled `trailofbits/algo:latest`. These values can be changed in the `Makefile` itself.
+2. Run `make docker-deploy`. By default this will use the given copy of `config.cfg` and populate the project's root `~/configs/` directory with your client configuration data. If you would rather use a local directory and `config.cfg` file, then follow steps 2 and 3 before changing the `CONFIGURATIONS` variable in the `Makefile`.
```
Comment (TC1977) [inline docs/deploy-from-docker.md]: Please clarify the "steps 2 and 3" that are referred to here?

Comment (johnwesley) [reply docs/deploy-from-docker.md]: Yes, I will expand on this.

**Comment by johnwesley** [comment]
@jackivanov conflicts have been resolved and the first local test `docker-ci-local` passes. Upgrade users test is failing on a step.

**Comment by johnwesley** [comment]
@jackivanov I've also tested `algo` locally with both Docker and `virtualenv` deployments.

**Comment by jackivanov** [comment]
Do we really need to mark the image as `trailofbits/algo:latest`? Won't we get conflicts with [the one from dockerhub](https://cloud.docker.com/u/trailofbits/repository/docker/trailofbits/algo) in some cases?

**Comment by johnwesley** [comment]
No we don't. Would you rather use a release tag of some sort?
We can use `$(IMAGE):$(TAG)` in the test runs and set the tag top-scope same as we're doing for `docker-build` and `docker-deploy`. Another option is to use a git tag.
Example:
`TAG := $(shell git describe --tags)`

`Successfully tagged trailofbits/algo:v1.0-65-g91a44bf`

**Comment by jackivanov** [comment]
I think we're good after you complete [this](https://github.com/trailofbits/algo/pull/1341/files#r290357912)

**Comment by johnwesley** [comment]
Sorry for this, but I think attempting to remove Python virtualenv from the docker deploy is a bad idea at this point and will cause systemic problems down the line. It's too ingrained into the app and I'm not 100% confident this will work for every situation presented. I say we move forward with the `Makefile` for Docker and tackle `virtualenv` in another PR.

**Comment by jackivanov** [comment]
@johnwesley I agree. Could you revert the changes, please?

**Comment by jackivanov** [comment]
The tests are failing now @johnwesley

**Comment by johnwesley** [comment]
@jackivanov I am aware and will try to fix resolve them today.

**Comment by johnwesley** [comment]
@jackivanov Docs have been updated and tests pass.

**Comment by johnwesley** [comment]
Can we just close this PR so I can open a new one with just the `Makefile` addition for Docker and tests?

**Pull request closed without merging**